[move 2024] Method syntax docs (#16760)
MystenLabs/sui: external-crates/move/documentation/book/src/method-syntax.md

Additionally, this can make code more readable, as it reduces the amount of boilerplate needed to
call a function and makes it easier to chain function calls.

## Syntax

The syntax for calling a method is as follows:

```text
<expression> . <identifier> <[type_arguments],*> ( <arguments> )
```

For example

```move
coin.value();
*nums.borrow_mut(i) = 5;
```

## Method Resolution

When a method is called, the compiler will statically determine which function is called based on
the type of the receiver (the argument on the left-hand side of the `.`). The compiler maintains a
mapping from type and method name to the module and function name that should be called. This
mapping is created fom the `use fun` aliases that are currently in scope, and from the appropriate
functions in the receiver type's defining module. In all cases, the receiver type is the first
argument to the function, whether by-value or by-reference.

In this section, when we say a method "resolves" to a function, we mean that the compiler will
statically replace the method with a normal [function](./functions.md) call. For example if we have

> `x.foo(e)` with foo resolving to `a::m::foo`, the compiler will replace `x.foo(e)` with

Reviewer: ```suggestion
`x.foo(e)` with `foo` resolving to `a::m::foo`, the compiler will replace `x.foo(e)` with
```

`a::m::foo(x, e)`, potentially [automatically borrowing](#automatic-borrowing) `x`.

### Functions in the Defining Module

In a type’s defining module, the compiler will automatically create a method alias for any function
declaration for its types when the type is the first argument in the function. For example,

```move
module a::m {
    public struct X() has copy, drop, store;
    public fun foo(x: &X) { ... }
    public fun bar(flag: bool, x: &X) { ... }
}
```

The function `foo` can be called as a method on a value of type `X`. However, not the first argument
(and one is not created for `bool` since `bool` is not defined in that module). For example,

```move
fun example(x: a::m::X) {
    x.foo(); // valid
    // x.bar(true); ERROR!
}
```

### `use fun` Aliases

Like a traditional [`use`](uses.md), a `use fun` statement creates an alias local to its current
scope. This could be for the current module or the current expression block. However, the alias is
associated to a type.

The syntax for a `use fun` statement is as follows:

```move
use fun <function> as <type>.<method alias>;
```

> This creates an alias for the `<function>` which can be received by the `<type>` as

Reviewer: ```suggestion
This creates an alias for the `<function>`, which the `<type>` can receive as
```

`<method alias>`.

For example

```move
module a::cup {
    public struct Cup<T>(T) has copy, drop, store;

    public fun cup_borrow<T>(c: &Cup<T>): &T {
        &c.0
    }

    public fun cup_value<T>(c: Cup<T>): T {
        let Cup(t) = c;
        t
    }

    public fun cup_swap<T: drop>(c: &mut Cup<T>, t: T) {
        c.0 = t;
    }
}
```

We can now create `use fun` aliases to these functions

```move
module b::example {
    use fun a::cup::cup_borrow as Cup.borrow;
    use fun a::cup::cup_value as Cup.value;
    use fun a::cup::cup_swap as Cup.set;

    fun example(c: &mut Cup<u64>) {
        let _ = c.borrow(); // resolves to a::cup::cup_borrow
        let v = c.value(); // resolves to a::cup::cup_value
        c.set(v * 2); // resolves to a::cup::cup_swap
    }
}
```

Note that the `<function>` in the `use fun` does not have to be a fully resolved path, and an alias
can be used instead, so the declarations in the above example could equivalently be written as

```move
    use a::cup::{Self, cup_swap};

    use fun cup::cup_borrow as Cup.borrow;
    use fun cup::cup_value as Cup.value;
    use fun cup_swap as Cup.set;
```

> While these examples are cute for renaming the functions in the current module, the feature is

Reviewer: aww, that is cute!

perhaps more useful for declaring methods on types from other modules. For example, if we wanted to

> add a new utility to `Cup` we could do so with a `use fun` alias and still use method syntax

Reviewer: ```suggestion
add a new utility to `Cup`, we could do so with a `use fun` alias and still use method syntax
```

```move
module b::example {

    fun double(c: &Cup<u64>): Cup<u64> {
        let v = c.value();
        Cup::new(v * 2)
    }

}
```

> Normally, we would be stuck having to call it as `double(&c)` since `b::example` did not define

Reviewer: ```suggestion
Normally, we would be stuck having to call it as `double(&c)` because `b::example` did not define
```
nit: 'since' has both a causal and temporal meaning. Because of this, because is usually preferred.

`Cup`, but instead we can use a `use fun` alias

```move
    fun double_double(c: Cup<u64>): (Cup<u64>, Cup<u64>) {
        use fun b::example::double as Cup.dub;
        (c.dub(), c.dub()) // resolves to b::example::double in both calls
    }
```

> While a `use fun` can be made in any scope, the target `<function>` of the `use fun` must have a

Reviewer: ```suggestion
While `use fun` can be made in any scope, the target `<function>` of the `use fun` must have a
```

to be used outside of its declared scope. A `use fun` can be made `public` if it is declared in the
module that defines the receivers type, much like the method aliases that are
[automatically created](#functions-in-the-defining-module) for functions in the defining module. Or
conversely, one can think that an implicit `public use fun` is created automatically for every
function in the defining module that has a first argument of the receiver type (if it is defined in
that module). Both of these views are equivalent.

```move
module a::cup {
    public struct Cup<T>(T) has copy, drop, store;

    public use fun cup_borrow as Cup.borrow;
    public fun cup_borrow<T>(c: &Cup<T>): &T {
        &c.0
    }
}
```

In this example, a public method alias is created for `a::cup::Cup.borrow` and
`a::cup::Cup.cup_borrow`. Both resolve to `a::cup::cup_borrow`. And both are "public" in the sense
that they can be used outside of `a::cup`, without an additional `use` or `use fun`.

```move
module b::example {

    fun example<T: drop>(c: a::cup::Cup<u64>) {
        c.borrow(); // resolves to a::cup::cup_borrow
        c.cup_borrow(); // resolves to a::cup::cup_borrow
    }
}
```

> `public use fun`s thus serve as a way of renaming a function if you need wish to give it a cleaner

Reviewer: ```suggestion
The `public use fun` declarations thus serve as a way of renaming a function if you want to give it a cleaner
```
Not sure if this is the right verbiage, but "`public use fun`s"  just looks kinda weird.

name for use with method syntax. This is especially helpful if you have a module with multiple
types, and similarly named functions for each type.

```move
module a::shapes {

    public struct Rectangle { base: u64, height: u64 }
    public struct Box { base: u64, height: u64, depth: u64 }

    // Rectangle and Box can have methods with the same name

    public use fun rectangle_base as Rectangle.base;
    public fun rectangle_base(rectangle: &Rectangle): u64 {
        rectangle.base
    }

    public use fun box_base as Box.base;
    public fun box_base(box: &Box): u64 {
        box.base
    }

}
```

> Another use for `public use fun`s is adding methods to types from other modules. This can be helpful

Reviewer: ```suggestion
Another use for `public use fun` is adding methods to types from other modules. This can be helpful
```

```move
module a::cup {
    public struct Cup<T>(T) has copy, drop, store;

    public fun new<T>(t: T): Cup<T> { Cup(t) }
    public fun borrow<T>(c: &Cup<T>): &T {
        &c.0
    }
    // `public use fun` to a function defined in another module
    public use fun a::utils::split as Cup.split;
}

module a::utils {
    use a::m::{Self, Cup};

    public fun split<u64>(c: Cup<u64>): (Cup<u64>, Cup<u64>) {
        let Cup(t) = c;
        let half = t / 2;
        let rem = if (t > 0) t - half else 0;
        (cup::new(half), cup::new(rem))
    }

}
```

And note that this `public use fun` does not create a circular dependency, as the `use fun` is not
present after the module is compiled--all methods are resolved statically.

### Interactions with `use` Aliases

A small detail to note is that method aliases respect normal `use` aliases.

```move
module a::cup {
    public struct Cup<T>(T) has copy, drop, store;

    public fun cup_borrow<T>(c: &Cup<T>): &T {
        &c.0
    }
}

module b::other {
    use a::cup::{Cup, cup_borrow as borrow};

    fun example(c: &Cup<u64>) {
        c.borrow(); // resolves to a::cup::cup_borrow
    }
}
```

> A potentially helpful way to think about this is that a `use` creates an implicit `use fun` alias

Reviewer: ```suggestion
A helpful way to think about this is that `use` creates an implicit `use fun` alias
```

for the function whenever it can. In this case the `use a::cup::cup_borrow as borrow` creates an

> implicit `use fun a::cup::cup_borrow as Cup.borrow` since it would be a valid `use fun` alias. Both

Reviewer: ```suggestion
implicit `use fun a::cup::cup_borrow as Cup.borrow` because it would be a valid `use fun` alias. Both
```

> views are equivalent. This line of reasoning can help to how specific methods will resolve with

Reviewer: ```suggestion
views are equivalent. This line of reasoning can inform how specific methods will resolve with
```
or "help to **show** how"

module a::m {
    public struct X() has copy, drop;
    public fun by_val(_: X) {}
    public fun by_ref(_: &X) {}
    public fun by_mut(_: &mut X) {}

    public struct Y has drop { x: X }

    fun example(mut y: Y) {
        y.x.by_ref(); // resolves to a::m::by_ref(&y.x)
        y.x.by_mut(); // resolves to a::m::by_mut(&mut y.x)
    }
}
```

Note that in both examples, the local variable had to be labeled as [`mut`](./variables.md) to allow
for the `&mut` borrow. Without this, there would be an error saying that `x` (or `y` in the second
example) is not mutable.

Keep in mind that without a reference, normal rules for variable and field access come into play.
Meaning a value might be moved or copied if it is not borrowed.

```move
module a::m {
    public struct X() has copy, drop;
    public fun by_val(_: X) {}
    public fun by_ref(_: &X) {}
    public fun by_mut(_: &mut X) {}

    public struct Y has drop { x: X }
    public fun drop_y(y: Y) { y }

    fun example(y: Y) {
        y.x.by_val(); // copies `y.x` since `by_val` is by-value and `X` has `copy`
        y.drop_y(); // moves `y` since `drop_y` is by-value and `Y` does _not_ have `copy`
    }
}
```

## Chaining

> Method calls can be chained, since the receiver of the method can be any expression.

Reviewer: ```suggestion
Method calls can be chained, because any expression can be the receiver of the method.
```

```move
module a::shapes {
    public struct Point has copy, drop, store { x: u64, y: u64 }
    public struct Line has copy, drop, store { start: Point, end: Point }

    public fun x(p: &Point): u64 { p.x }
    public fun y(p: &Point): u64 { p.y }

    public fun start(l: &Line): &Point { &l.start }
    public fun end(l: &Line): &Point { &l.end }

}

module b::example {
    use a::shapes::Line;

    public fun x_values(l: Line): (u64, u64) {
        (l.start().x(), l.end().x())
    }

}
```

> In this example, `l.start().x()` the compiler first resovles `l.start()` to `a::shapes::start(&l)`.

Reviewer: ```suggestion
In this example, `l.start().x()` the compiler first resolves `l.start()` to `a::shapes::start(&l)`.
```
I think this sentence is missing a word or three somewhere around `l.start().x()`?

> Then `.x()` is resolved to `a::shapes::x(a::shapes::start(&l))`. Similarly, `l.end().x()`. Keep in

Reviewer: I think a few more words around l.end().x() might also help.

mind, this feature is not "special"--the left-hand side of the `.` can be any expression, and the
compiler will resolve the method call as normal. We simply draw attention to this sort of "chaining"

> since it is a common practice to increase readability.

Reviewer: ```suggestion
because it is a common practice to increase readability.
```
Sorry, but I can't stop now...
